File: cmd/ui/src/api/dlt/hedera.ts
export const getHederaMessage = async (
  lastSequenceNumber: number,
): Promise<string> => {
  const TOPIC_ID = "0.0.1033";
  const MIRROR_NODE_BASE_URL = "http://10.18.1.35:5551";
  try {
    const response = await fetch(
      `${MIRROR_NODE_BASE_URL}/api/v1/topics/${TOPIC_ID}/messages?limit=1&order=asc&sequencenumber=${lastSequenceNumber}`,
    );

    if (!response.ok) {
      const error = await response.text();
      throw new Error(`HTTP error ${response.status}: ${error}`);
    }

    const chunks = (await response.json())["messages"];
    const messages = [];
    for (const chunk of chunks) {
      const message = chunk["message"];
      messages.push(atob(message));
    }
    return messages[0];
  } catch (error) {
    console.log(error);
    return "";
  }
};

File: internal/dlt/hedera.go
package dlt

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"time"

	"clever.eu/dashboard/internal/config"
	hedera "github.com/hiero-ledger/hiero-sdk-go/v2/sdk"
)

type HederaConsensusDLTClient struct {
	cfg    config.HederaConfig
	logger *slog.Logger
	client *hedera.Client
}

func NewHederaConsensusDLTClient(
	cfg config.HederaConfig,
	logger *slog.Logger,
) *HederaConsensusDLTClient {
	return &HederaConsensusDLTClient{cfg: cfg, logger: logger}
}

func (c *HederaConsensusDLTClient) GetMessages() ([]string, error) {
	yesterday := time.Now().AddDate(0, 0, -7).Unix()
	messages := []string{}
	for _, topic := range c.cfg.Topics {
		response, err := http.Get(fmt.Sprintf(
			"%s/api/v1/topics/%s/messages?order=desc&timestamp=gt:%d",
			c.cfg.Mirror.GetURL(),
			topic,
			yesterday),
		)
		if err != nil {
			return []string{}, err
		}

		body, _ := io.ReadAll(response.Body)
		var msgs struct {
			Messages []map[string]any `json:"messages,omitempty"`
		}
		err = json.Unmarshal(body, &msgs)
		if err != nil {
			return []string{}, err
		}

		// Messages received in decending order because the default
		// limit is 25. While we could increase the limit, if recent
		// messages are more than the limit we will not get them in
		// the response
		//
		// Therefore, we fetch in descending order, but in order to return
		// the messages in ascending order, we reverse them here.
		ascendingOrderMessages := []map[string]any{}
		for i := range msgs.Messages {
			ascendingOrderMessages = append(
				ascendingOrderMessages,
				msgs.Messages[len(msgs.Messages)-1-i],
			)
		}

		for _, msg := range ascendingOrderMessages {
			encodedValue := msg["message"].(string)
			value, err := base64.StdEncoding.DecodeString(encodedValue)
			if err != nil {
				return []string{}, err
			}
			messages = append(messages, string(value))
		}

	}
	return messages, nil
}
